mantidproject/mantid
# Imaging: Refactored tool dialogues and settings to be ready for future extensions

Part of Imaging functionality and GUI for IMAT Features Umbrella #17809 
All future refactoring work will be based on this PR and cannot be merged before it.
## Description of work.
- All dialogues are now handled through polymorphism, `TomoToolConfigDialogBase` is now base class for the tools, providing common virtual methods.
- All the dialogues files and the classes follow the naming pattern: `TomoToolConfigDialog<ToolName>`
- All settings now use the common base class they have `TomoRecToolConfig` instead of being passed around in a struct.
- Following those changes command line argument building in TomographyIfaceModel.cpp is now more straightforward. The methods for command line have been updated to make use of the polymorphic dialogues and settings 
- The tool settings configuration has been moved out in the presenter and the logic is handled there. The View now only executes the dialogue (the execute function also sets up the GUI parts).
- The tests for the view/presenter/model have been updated. I believe more tests could be added, but that would be in a separate issue (suggestions are welcome).

**To test:**
- Build and run the interface.
- Go through [the documentation examples](http://docs.mantidproject.org/nightly/interfaces/Tomographic_Reconstruction.html), ALL functionality should have been retained.
- Very very thorough code review as I suspect many a thing can be improved.
- Any feedback for the interface usability/structure would also be welcome.
- To run remote reconstructions you'll need an account on SCARF, local reconstructions are currently _hard coded_ to only work on NDL-IMAT, so reconstruct will just show the command output, verifying that it is correct (the same as previous Mantid versions, release 3.8 or master) 



Fixes #17244 .



---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
